pkg/pool-weighted/contracts/lbp/LBPMigrationRouter.sol

```diff
@@ -0,0 +1,199 @@
+
+    modifier onlyLBPOwner(ILBPool lbp) {
+        {
+            address lbpOwner = Ownable(address(lbp)).owner();
+            if (msg.sender != lbpOwner) {
+                revert SenderIsNotLBPOwner();
+            }
+        }
+
+        _;
+    }
+
+    constructor(
+        BalancerContractRegistry contractRegistry,
+        string memory version
+    ) Version(version) VaultGuard(contractRegistry.getVault()) {
+        (address weightedPoolFactoryAddress, bool isWeightedPoolFactoryActive) = contractRegistry.getBalancerContract(
+            ContractType.POOL_FACTORY,
+            "WeightedPool"
+        );
+        if (!isWeightedPoolFactoryActive) {
+            revert ContractIsNotActiveInRegistry("WeightedPool");
+        }
+
+        _weightedPoolFactory = WeightedPoolFactory(weightedPoolFactoryAddress);
+    }
+
+    /// @inheritdoc ILBPMigrationRouter
+    function migrateLiquidity(
+        ILBPool lbp,
+        uint256[] memory exactAmountsIn,
+        uint256 minAddBptAmountOut,
+        uint256[] memory minRemoveAmountsOut,
+        address excessReceiver,
+        WeightedPoolParams memory params
+    ) external onlyLBPOwner(lbp) returns (IWeightedPool, uint256) {
+        return
+            _migrateLiquidity(
+                lbp,
+                exactAmountsIn,
+                minAddBptAmountOut,
+                minRemoveAmountsOut,
+                msg.sender,
+                excessReceiver,
+                params,
+                false
+            );
+    }
+
+    /// @inheritdoc ILBPMigrationRouter
+    function queryMigrateLiquidity(
+        ILBPool lbp,
+        uint256[] memory exactAmountsIn,
+        address sender,
+        address excessReceiver,
+        WeightedPoolParams memory params
+    ) external returns (uint256 bptAmountOut) {
+        (, bptAmountOut) = _migrateLiquidity(
+            lbp,
+            exactAmountsIn,
+            0,
+            new uint256[](exactAmountsIn.length),
+            sender,
+            excessReceiver,
+            params,
+            true
+        );
+    }
+
+    function migrateLiquidityHook(MigrationHookParams memory params) external onlyVault returns (uint256 bptAmountOut) {
+        (, uint256[] memory removeAmountsOut, ) = _vault.removeLiquidity(
+            RemoveLiquidityParams({
+                pool: address(params.lbp),
+                from: params.sender,
+                maxBptAmountIn: IERC20(address(params.lbp)).balanceOf(params.sender),
+                minAmountsOut: params.minRemoveAmountsOut,
+                kind: RemoveLiquidityKind.PROPORTIONAL,
+                userData: new bytes(0)
+            })
+        );
+
+        for (uint256 i = 0; i < removeAmountsOut.length; i++) {
+            if (params.exactAmountsIn[i] > removeAmountsOut[i]) {
+                revert InsufficientInputAmount(params.tokens[i], removeAmountsOut[i]);
+            }
+
+            uint256 remainingBalance = removeAmountsOut[i] - params.exactAmountsIn[i];
```

Comment:
Here we'll need to apply the percentage considering the new pool's weight, and computing the input amounts accordingly to preserve the price considering the difference between the LBP end weights and weighted pool's starting weights.